### PR TITLE
Dashboard Functionality

### LINKED ISSUE #9: Add new dashboard functionality
These task is relatively open ended, but should be used as a guide. When referring to the 'dashboard', we are referring to the first page in scout a user sees after logging in. Today, all that is shown is device counts for all of the platforms scout manages, plus a count of the servers. We'd like to have a configurable grid of items with useful information below these counts. Some ideas of what that could be include: 

1. Viewing the counts of devices in saved reports (Required for PR) 
2. View health check information from various servers
3. View the count of users in scout 
4. View scout server information 
5. View frequency of cron jobs executions 
6. View scout total device count over time 

**Tasks:**
1. Add a new database column to the reports called 'show_in_dashboard' that is of boolean  type
2. In the report create / edit view allow the user to check this new box to show this saved report in the dashboard 
3. Import a charts library like [chart.js](https://www.chartjs.org/) to be used in dashboard visualizations. 
4. Add a new endpoint to get the saved reports and their results to load into the dashboard 
5. Upon page load, show some charts of the saved reports in the dashboard. 
6. Implement any of the other ideas above. 

**PR Ready When:**
1. Reports can be created / saved with the 'show in dashboard' checkbox checked
2. On the dashboard you are able to view the device counts in the various reports 
3. BONUS: One or more of the other ideas

### PR DESCRIPTION
**Pull request by alex-stout**
## Changelog
- Add radio box to toggle a report being visible from the dashboard
- Added styling for cards on the dashboard
- Cards Added:
    - Graph including total count of devices in each report
    - Total count of users (only visible with permission to edit users)
- To enable the selection/toggling of reports in the dashboard, a column was added to the SQL schema
- API endpoint was created to better facilitate the count of the total devices in a report. This was done so that it is more efficient and happens all on the server. Without this endpoint, the calculations took a *very* long time. Now it's all in memory and is pretty quick.

## Screenshots
Screenshot from 2020-02-14 16-47-43
Screenshot from 2020-02-14 16-46-21

## Contributions
@fostermorgan

Resolves #9

### CODE REVIEW

File: api/controllers/reports.js [hunk 1]
```diff
@@ -49,6 +49,27 @@ reports.get('/', function(req,res){
...
+ * @returns {object} 200 - A List of all reports
+ * @returns {Error}  500 - Unable to get reports
+ */
+reports.get('/dashboard', function(req,res){
+  report.getReportsToShowInDashboard()
+  .then(function(reports){
+    return res.status(200).send(reports);
+  })
+  .catch(error => {
+    console.log(error);
```
Comment (jacobschultz) [inline api/controllers/reports.js]: Do we need this `console.log` if we are writing to a log file?

Comment (jacobschultz) [reply api/controllers/reports.js]: Looks like there a few console logs we could clean up I think

File: api/init_sql/scout.sql [hunk 2]
```diff
@@ -72,6 +72,7 @@ CREATE TABLE `reports` (
   `conditions_count` int(11) DEFAULT NULL,
   `type` varchar(128) NOT NULL DEFAULT '',
   `fields_to_select` longtext,
+  `show_in_dashboard` boolean NOT NULL DEFAULT 0,
```
Comment (jacobschultz) [inline api/init_sql/scout.sql]: Not a big deal, but I think the rest the bool fields use `tinyint(1)` - should end up the same though

File: app/js/scripts.js [hunk 3]
```diff
@@ -134,148 +134,239 @@ function editReportById(reportId){
...
+      console.log(xhr);
+    })
+}
+
+async function calculateReportResultCount() {
+  var reportObject = await asyncGetRequestObject('/reports/dashboard/', null, 'GET'); //gets all reports in mySQL
+  var outputObj = reportObject;
+  var reportCalls = [];
+  for (var i = 0; i < reportObject.length; i++) { //for each report
+    reportCalls = await asyncGetRequestObject('/reports/' + reportObject[i].id + '/count', null, 'GET');
```
Comment (jacobschultz) [inline app/js/scripts.js]: You could consider using Promise.all() here to kick these off all at the same time and not block in the loop

File: app/js/scripts.js [hunk 4]
```diff
@@ -134,148 +134,239 @@ function editReportById(reportId){
...
+    loadAllUsers().then(data => {
+      $('#userCountCard').show();
+      $('#userCard').text(data.length)
+    }).catch(() => {
+      $('#userCountCard').hide();
+    })
+}
+
+async function loadAllUsers() {
+  var users = await asyncGetRequestObject('/users/all', null, 'GET');
```
Comment (jacobschultz) [inline app/js/scripts.js]: could do this in one line `return await asyncGetRequestObject('/users/all', null, 'GET');`

File: app/js/scripts.js [hunk 5]
```diff
@@ -1522,46 +1616,48 @@ function changeReportView(deviceType, operation){
...
-      if (!("auth_token" in sessionStorage)){
+      if (!("auth_token" in sessionStorage)) {
         $('#login-user-modal').modal('show');
       }
     })
   } else {
     renderPage();
   }
 
+
```
Comment (jacobschultz) [inline app/js/scripts.js:1672]: This is a pretty big diff - I assume this just went through an auto formatter? Was any regression testing performed to make sure this didn't break anything? Should probably be fine, just want to double check.

Comment (alex-stout) [reply app/js/scripts.js:1672]: It did. I'll revert it if you prefer.

**Comment by alex-stout** [comment]
I’m going to check compatibility with current systems to make sure that the DB schema change won’t break existing systems that don’t upgrade.

We haven’t tested this test.

**Comment by alex-stout** [comment]
Just tested the case of not updating and it causes some problems. I added a pop-up to let the user know that that might be an issue. It should only happen when they upgrade.

Here's what it looks like: 
Screen Shot 2020-02-26 at 9 05 10 AM

_The trigger for the notification is if the API returns a 500 error._

File: app/js/scripts.js (hunk 5 above)
Comment (jacobschultz) [reply app/js/scripts.js:1672]: Nope, that's okay. Just wanted to double check. Probably makes it cleaner in the long run anyway.

**Comment by jacobschultz** [comment]
This looks good to me! I like the solution of checking the database for the proper fields. We didn't really have anything else like that before, so it would just break if the databse wasn't up to date. This gives the user some feedback.

**Pull request merged**